### ISSUE TITLE
[BUG]  onInit() is executed AFTER the widget renders

### ISSUE THREAD

**Issue by alexsmartens** [thingsboard/thingsboard#1517]
`self.onInit()` is executed only after the widget renders, which is a problem for initialization of the initial values. Is there any way around it?

An analog of the method which I am looking for is [componentWillMount](https://developmentarc.gitbooks.io/react-indepth/content/life_cycle/birth/premounting_with_componentwillmount.html) in React. This method is only called before the initial render.

### Code to replicate the issue:
HTML:
```
<script>
    console.log($customValue)
</script>
```
JS:
```
 self.onInit = function() {
        self.ctx.$scope.data = self.ctx.defaultSubscription.data
        $customValue = self.ctx.defaultSubscription.data[0]

        console.log($customValue)
    }
```

### Expected result:
`$customValue` is printed two times in the console 

### Actual result:
- first printed line:
`ReferenceError: $customValue is not defined[Learn More]`
- second printed line:
actual value of `$customValue`

**Comment by nsantos111** [comment]
@alexmartens I have the same issue occurring randomly on a widget. Have you figured out how to handle it? Thanks.

**Comment by alexsmartens** [comment]
Hi @nsantos111,
Yes, I have. As the first step, I stopped writiting the script in the html part and utilized `onInit`,`onDataUpdated` and other default (this might be a good enough solution).

Then, I needed more flexibility and started using [d3](https://d3js.org/) instead of the default thingsboard library. In this case you can definitely control everything what happens because you control the initialization.

**Comment by alexsmartens** [comment]
I hope it helps. If it doesn't then post more information about your problem and I'll try to help you

**Comment by alexsmartens** [comment]
@mp-loki, @luPhz,
Any suggestions?

**Comment by nsantos111** [comment]
Thanks for your reply.

I already don't have the script in the html part and I'm using `onInit()`.

My issue is that I'm using `ng-indeterminate` and `ng-checked` with functions that check some properties of the objects of an array that is declared and initialized at the beginning of `onInit()`. 
When refreshing the dashboard page I get occasional `Cannot read property '...' of undefined` errors on the html rendering.

**Comment by alexsmartens** [comment]
Ok, let's see if I got the problem right: you are adding checkboxes to the UI. Then, onInit() you look at whether the checkboxes are selected or not. At this step, you sometimes get the error.

If this is the case, try adding the checkboxes inside the init() instead of declaring them in the html part of the widget. In this way you will always have them before you look at their state. You can do that by using jQuery (it's already declared in TB).

If I did not get it right, please give more detailed description of the problem

**Comment by alexsmartens** [comment]
One more alternative, you can use a promise with a delay of say 100ms before running everything inside your init() and run that part of the init() only when this promise is executed. It's a dirty solution but it's probably the most simplest one

**Comment by nsantos111** [comment]
Yes, that's more or less the case.  I'm not looking for whether the checkboxes are selected or not, but for some properties of the array objects - and have one checkbox row for each object. The array is filled in at `onInit()` beginning. The issue occurs when the array is not filled before the rendering of the html which only happens sometimes.

I understood your suggestion (of _building_ the elements on `onInit` instead of in the html part) and if there's no other solution I'll give it a try as soon as I finish other widget I'm implementing - Never used jQuery before...

Thanks

**Comment by nsantos111** [comment]
Hi @alexsmartens,

After reading some articles about jQuery I put my hands on the task of moving from my very complex static html to the jQuery built layout. During the process of wall breaking with my head, I found an error on my initial JS code. After fixing it, the errors on the functions called from the html were fixed - the issue was not related with `onInit()` not being called before the html rendering after all!


Thanks for your support and for pointing the jQuery option – is something that I must get into but definitely with a more simple html layout.
